### PR TITLE
Describing fast/standard review

### PR DESCRIPTION
**Pull request by jtulach**
Introducing trivial/complex/no API change and related reviews: fast-track/standard one.

### CODE REVIEW

**Comment by jtulach** [comment]
@chumer: Do all project changes to the public API count as "API changes"? I am thinking of the processor project which has public API but it is not "really" API. I guess also test projects are excluded from this rule right?

@jtulach: The "exported packages" are subject to API compatibility constraints. Right now those include the [ones visible in Javadoc](http://graalvm.github.io/truffle/javadoc/0.11/). The DSL Processor isn't there and is considered to be in an implementation package. Also the tests (except [TCK](http://graalvm.github.io/truffle/javadoc/0.11/com/oracle/truffle/tck/package-summary.html)) are  internal, as they are not part of any distribution.

**Comment by woess** [comment]
This document only talks about API reviews. Are you implying non-API changes don't need a review?

**Comment by jtulach** [comment]
Is it better [now](9c1349f1d183cb034b7506f4910776afa59be3af)?

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -3,10 +3,79 @@ There is a white-list for contributors who have signed the OCA so please add a c
...
+do so, it is important to stick with essential properties of a good API change:
+
+- be backward compatible
+- be well documented
+- have sufficient test coverage
+- be ready for future evolution
+
+If your change satisfies these properties you are eligible for a **fast-track** review: create a pull request,
+let people know about it, assign a responsible reviewer, react to *requests for change* and after a week 
+you should be ready to merge. Announce the *final 24 hour count down* and then merge or convince an
```
Comment (woess) [inline CONTRIBUTING.md]: Do we really need a *24 hour count down* thing?

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -3,10 +3,79 @@ There is a white-list for contributors who have signed the OCA so please add a c
...
+No API Change
+-------------
+
+Adding tests verifying existing behavior, improving documentation to describe current behavior, refactoring 
+internal code without influencing functionality, mangling with private or package private members of API
+classes and other *internal changes* do not need an API review at all.
+
+It is still preferrable to designate additional reviewer to consider the intentions from a different 
+angle and merge the change. Alternatively one can treat these "no API changes" as trivial API changes
+and request **fast-track** review "just to be sure".
```
Comment (woess) [inline CONTRIBUTING.md]: This part isn't clear enough about if and how many reviewer such a change needs ("no API review at all", "*additional* reviewer"; if it's not fast-track by default, what is it then?).

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -3,10 +3,79 @@ There is a white-list for contributors who have signed the OCA so please add a c
 to your first pull request indicating the name you used to sign the OCA if it isn't clear
 from your github profile.
 
-Pull requests can only be merged by committers and a committer cannot merge
-his or her own pull requests.
+It is desirable to review all changes by at least one other person. When creating a pull request,
+make sure you designate an assignee to ensure the request is taken care of in a timely manner.
+The assignee will guide you trough the process and find the most suitable review style for your
```
Comment (woess) [inline CONTRIBUTING.md]: **through**

File: CONTRIBUTING.md (hunk 1 above)
Comment (ansalond) [inline CONTRIBUTING.md]: > after a week you should be ready to merge. Announce the *final 24 hour count down*
I would propose to omit detailed information about how long the procedure is expected to take and simply ask to wait for the assignee to accept the pull request. With the current procedure, if the assignee approves the request within, e.g., 2 hours, do you still have to wait 24 hours before merging? If instead the assignee is out-of-office, can the pull request be merged without any feedback?

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -3,10 +3,79 @@ There is a white-list for contributors who have signed the OCA so please add a c
...
+Complex API Change
+------------------
+
+If an API change is inherently backward *incompatible*, or if it is huge, it is subject to **standard** review.
+The purpose of such review is to seek wider consensus among involved parties. The ideal state of such review
+should:
+
+- describe the intention, use-cases and let everyone comment
+- nominate four "voting" reviewers
+- have an *inception call* where reviewers approve/reject the project and set constraints/directions
```
Comment (ansalond) [inline CONTRIBUTING.md]: What do you mean by "inception call"?

File: CONTRIBUTING.md (hunk 1 above)
Comment (jtulach) [inline CONTRIBUTING.md]: The goal isn't to speed things up, but give everyone a chance to participate. API changes are serious enough to require/allow participation of more people than a reporter and a single assignee.

NetBeans experience seems to indicate a week review period is usually enough for everyone to take a look. The 24h countdown serves as a last warning to wake up those who are slower and give them last chance to catch up.

Please note that in case of **fast-track** review proposed by a commiter, the assignee isn't really needed. Smooth **fast-track** review is a self-service. However proposing changes behind somebody back isn't particularly nice habit and shouldn't be encouraged.

Comment (woess) [inline CONTRIBUTING.md]: I don't understand what you mean by "smooth fast-track review is a self-service".

**Comment by woess** [comment]
This is `CONTRIBUTING.md` and the proposed change makes it too focused on reviews when actually it should be concerned with Contributing Guidelines. I'd actually suggest to start from scratch and take ideas from popular OSS projects.

Some ideas what should be in the document:
* Legal aspects: OCA. The license your contribution will be subject to, etc. (IANAL)
* Conventions: How to properly formulate your contribution and pull request. What to expect (reviews, comments, change/update requests). Code style, tests, CI, you name it.
* Technical aspects: e.g. Push changes to a branch in a fork repository and create a PR. How to sync your branch with master. How to use mx, run tests, ...
* Where to go when you have questions (mailing list, ...).

**Comment by jtulach** [comment]
OCA is covered in first paragraph. That is indeed essential to mention. How to "formulate a request" is described - it needs to be compatible, documented, tested and ready for future evolution. I don't consider formatting important too waste characters on it. The presence of mx is indeed an obstacle, but describing that is out of scope of contributing document - moreover it is in README. Description of how people should work with git or github can be found elsewhere. Mailing list is mentioned in README, but that is more for user questions - github style of proposing changes is to open a pull request.

**Comment by woess** [comment]
Right, some of that is already in README.md. Doesn't need to be mentioned here, then.
It's still relevant that one has to pass the gate checks, which includes checkstyle, formatting, findbugs, unittests, javadoc and signature tests. So it could be described here what to do when one of these checks failed (mx). Perhaps `mx gate` is enough to know, though.

**Comment by woess** [comment]
I'd like to merge in https://github.com/graalvm/graal-core/pull/36 because I think it's a reasonable policy and we should keep our policies consistent across repositories.

File: CONTRIBUTING.md (hunk 1 above)
Comment (woess) [inline CONTRIBUTING.md]: I understand that certain pull requests need to stay open long enough to give interested parties a chance to participate in the review but I don't like fixed time constraints.

File: README.md [hunk 5]
```diff
@@ -97,7 +97,12 @@ and then it is possible to include the artifacts as dependencies to a `pom.xml`:
 ## Contributing
 
 You can contact the Truffle developers at graal-dev@openjdk.java.net mailing
-list.
+list. To contribute a change, verify it using
+```bash
+$ mx gate
+```
+and start a [pull request](https://help.github.com/articles/using-pull-requests/).
+Detailed info can be found in [contributing document](CONTRIBUTING.md).
```
Comment (dougxc) [inline README.md]: This last self-referential sentence can probably be omitted.

Comment (woess) [inline README.md]: this is actually README.md

Comment (dougxc) [inline README.md]: Oops - missed that. Github review don't make the boundaries between files clear enough in my opinion.

In this case, there should be a "the" after "found in".

File: CONTRIBUTING.md (hunk 4 above)
Comment (jtulach) [inline CONTRIBUTING.md]: By "call" I mean for example a skype conversation between the submitter and the reviewers. By "inception" I mean the call/discussion that happens before one starts to implement the change. Only then it is possible to change directions, require completely new features, limit the scope of the change, etc.

Comment (woess) [inline CONTRIBUTING.md]: but then by your description it has to happen before the pull request, so it's irrelevant imho

**Comment by jtulach** [comment]
I've added **Applying** section similar to  graalvm/graal-core#36 - we need to create the `accept` label.

I've clarified that the complex change pull request shall contain documentation changes when submitted. Good catch on the circularity.

**Comment by smarr** [comment]
Reading the current state, I have a general question:

What's the intention of the `CONTRIBUTING.md`? Should this outline the whole, rather complex process, or should it encourage external participation/contributions?
The described process seems very heavy-weight and slow, which is probably designed to intentionally slow down the change rate in Truffle and improve stability. However, the complexity and length of the process seems to me as something will discourage external contributions.

If external contribution is a goal, I'd aim for a minimal set of instructions/requirements here, and detail the process in another file.

**Comment by dougxc** [comment]
@smarr the stability of Truffle API is important enough that a contributor (internal or external) should think twice about contributing if they feel the process of getting a change into the API is too onerous.

**Comment by woess** [comment]
Thanks for your input @smarr.
> If external contribution is a goal, I'd aim for a minimal set of instructions/requirements here, and detail the process in another file.
I agree.

**Comment by mlvdv** [comment]
+1 input from @smarr 
> If external contribution is a goal, I'd aim for a minimal set of instructions/requirements here, and detail the process in another file.

I don't disagree with the reply from @dougxc , but first impressions matter a lot. If we encourage and even inspire potential contributors enough, and if we declare a commitment to API stability, then the process will seem worthwhile when it is eventually encountered.

**Comment by dougxc** [comment]
Maybe I just haven't been often enough in the shoes of an external collaborator deciding whether to contribute to a project but I would not find it discouraging for a project's CONTRIBUTING.md file to be somewhat exhaustive. It indicates that the project is of high quality and cares deeply about quality contributions. If I'm just going to base my decision on a quick impression, then I wouldn't look further than README.md which Github displays prominently (for that reason, I'm guessing).

If someone is serious about contributing, I'd rather they be aware of what it entails up front than later get (publicly) frustrated about the process of getting a PR accepted.

**Comment by smarr** [comment]
My experience with open source projects is that large initial contributions are highly unlikely. It is much more likely that new people contribute small fixes for typos, fix a trivial bug that annoys them, and perhaps proposes a minor feature.

Here the simplicity of github: push change & press create-PR button, really lower the entry barrier. And it is that simplicity that makes a difference to get people engaged.

But, that's just my personal experience. And again, I very much understand the desire to have a process, and I also understand the desire to emphasize the maturity of the project with it.

So, from my perspective there is no clear answer, just a choice of tradeoffs.

**Comment by jtulach** [comment]
Seven days of review is over. Thanks for your feedback. I will integrate tomorrow in current state. According to rules described in this document, you can block such step by requesting standard review.

I am convinced we need written down description/outline of API review steps. When we decide to reject some contribution, this is the document we will reference to explain why the contribution isn't good enough.

No process (complex or simple) will increase external contributions. Github people know how to contribute: fork & create pull request. Then it is up to humans to help them and comply with our integration standards.

**Comment by woess** [comment]
I do not approve of this one week deadline with ensuing automatic integration of changes. (I thought that was already clear from the comments.) As such I request a more thorough "standard review".

**Comment by jtulach** [comment]
OK, then let's have a skype meeting on Monday. Everyone, please speak up here if you want to join. I propose to chat at 4pm CET and I nominate @woess, @dougxc, somebody with NetBeans experience (either Martin Entlicher or Jan Štola) and @chumer as voting reviewers.

**Comment by thomaswue** [comment]
@mbrantner @ldaynes Please take a look at this proposal.

**Comment by smarr** [comment]
I request to be part of this meeting as well. I believe this is important for the impression the project gives to potential academic collaborators.

**Comment by dougxc** [comment]
Unfortunately, I will not be available on Monday. Any other day and time works for me.

**Comment by jtulach** [comment]
Re. *"I thought that was already clear from the comments"* - no, it wasn't and that is the part of the problem. We need a way to separate comments from requests. There needs to be a clear way for everyone to express a request and take responsibility for that.

**Comment by entlicher** [comment]
In the Complex API Change section there is "organize phone or skype call..." I don't think we should limit the communication technologies in the document and permit people to agree on e.g. a Zoom conference if they want to. I'd rephrase it as "organize a call (e.g. phone, skype, etc.) where reviewers discuss..."

**Comment by entlicher** [comment]
Do we want to reference the CHANGELOG.md in this document? Should an entry be added into Changelog when an API change is made?
There is nothing about versioning. In NetBeans we increase the version number and add @since tags in the Javadoc of added API classes. Is it worth to implement a similar approach in Truffle as well? I'd propose the third digit of the version number be increased after an API change.

**Comment by entlicher** [comment]
Should we require that every newly contributed API be covered by tests?
Or would this impose too high contribution barrier?

**Comment by jtulach** [comment]
Yes, every API change has to be covered by tests (unless decided otherwise). The document already mentions that by saying "have sufficient test coverage".

**Comment by jtulach** [comment]
There is #42 to deal with API change list. It will require some work. Let's leave it out of scope of the process change for now.

**Comment by jtulach** [comment]
@thomaswue if you are OK with your latest version please add `accept` label and I merge.

**Comment by woess** [comment]
Can we just squash this to one commit, please?

**Comment by jtulach** [comment]
I can handle the squashing once `accept` is added by @thomaswue
